packages/synapse-react-client/src/components/useImmutableTableQuery.ts

```diff
@@ -148,56 +184,82 @@ export default function useImmutableTableQuery(
     return cloneDeep(initQueryRequest)
   }, [initQueryRequest])
 
+  const [commitAfterDebounce, setCommitAfterDebounce] = useState(false)
+
+  useDebouncedEffect(
+    () => {
+      if (commitAfterDebounce) {
+        commit()
+        setCommitAfterDebounce(false)
+      }
+    },
+    // nextQueryRequest MUST be included in the dependencies to ensure the debounce resets when it changes
+    [nextQueryRequest, commit, commitAfterDebounce],
+    DEBOUNCE_DELAY_MS,
+  )
+
   /**
    * Execute the given query request, updating all the data in the QueryContext to match the new query
    * @param {*} queryRequest Query request as specified by
    *                         https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/table/Query.html
    */
   const setQuery: ImmutableTableQueryResult['setQuery'] = useCallback(
-    (queryRequest): void => {
-      const newQueryRequest = cloneDeep(
+    (
+      queryRequest: React.SetStateAction<QueryBundleRequest>,
+      commitOptions?: QueryChangeCommitOptions,
+    ): void => {
+      const nextQueryRequest =
         typeof queryRequest === 'function'
-          ? queryRequest(getLastQueryRequest())
-          : queryRequest,
-      )
+          ? queryRequest(getNextQueryRequest())
+          : queryRequest
 
       // Remove null/empty array fields
-      lastQueryRequest.query = removeEmptyQueryParams(lastQueryRequest.query)
+      nextQueryRequest.query = removeEmptyQueryParams(nextQueryRequest.query)
 
-      setLastQueryRequest(newQueryRequest)
+      setNextQueryRequest(nextQueryRequest)
+      if (commitOptions) {
+        if ('debounce' in commitOptions && commitOptions.debounce) {
+          setCommitAfterDebounce(true)
+        }
+      } else {
+        commit()
```

Comment:
If there are `commitOptions` but `debounce` is either `undefined` or `false`, will it ever call `commit()`?

Reply:
Nope, but you would also get a compile-time error